### ISSUE TITLE
window DOM classes should be instanceof window.Function

### ISSUE THREAD

**Issue by klarkc** [tmpvar/jsdom#1737]
Hello, I found this issue when I was trying to use a CustomElement polyfill with jsdom. The polyfill validates element definition by checking instanceof passed classes.

It seems that all DOM HTML classes must be instanceof window.Function (at least my browsers): https://jsfiddle.net/0gxdxcpu/

With jsdom:

```javascript
jsdom.jsdom(
    '<script>console.log(window.HTMLElement instanceof window.Function)</script>',
    {virtualConsole: jsdom.createVirtualConsole().sendTo(console)}
);`
```

console output is false

**Comment by Sebmaster** [comment]
Sadly probably not fixable without making jsdom a loooot slower due to sandboxing 😞

**Comment by domenic** [comment]
We might be able to set window.Function to Node's global.Function. We've already done that for a few instanceof cases.

We should try to figure out some better solution for these instanceof related things though...

**Comment by klarkc** [comment]
Maybe we should let the user decides how will the window "global environment" be, with a parameter to do this?

**Comment by Sebmaster** [comment]
If you want to do that yourself, you can already, by overwriting `window.Function` in the `created` callback.

**Comment by klarkc** [comment]
@Sebmaster Hmmm you are right, but it's still not clear to the user that there is a "fake" Function there, using a parameter like `windowGlobalConstructor` might be more obvious that window "global" things is not node global things

**Comment by Sebmaster** [comment]
We'll still have to decide piece by piece which built-ins we want to overwrite with their node counterpart in the sandbox (like we're doing right now to get even basic stuff working). We'll start patching more and more stuff in there either way (until we actually create a whole new environment per sandbox, which is gonna be way too slow to be useful (probably)).

**Comment by klarkc** [comment]
@Sebmaster I gotcha. So there are more node things that was not overwritten and not in the docs? At least the common javascript things, like Function or Object, should be very well documented to not hide potential bugs like this one.

**Comment by domenic** [comment]
I don't think this is really a bug, it's just an unexpected consequence of having two different globals. A jsdom instance can be thought of like an iframe. In a browser you don't call it a bug or gotcha that `objectFromIFrame instanceof Object` is false.

Edit 2026-09-17: I'm not sure what my 2017 comment was trying to say exactly. The test code in the OP is something that works in browsers and doesn't work in jsdom. In an ideal world, we would make it work. https://github.com/jsdom/jsdom/issues/1207 and https://github.com/jsdom/jsdom/issues/2727 are related.

**Comment by klarkc** [comment]
Hmm thinking this way you are right... And for sake of the human being we should not mix node globals with internal jsdom globals. But that unexpected consequences that brings me concern, maybe it's more related to best practices and not to jsdom itself.
